src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/SystemModulesPlugin.java

```diff
@@ -587,440 +572,404 @@ private void dedups(ModuleDescriptor md) {
+        public byte[] getClassWriter(Configuration cf) {
+            return Classfile.build(classDesc,
+                    clb -> {
+                        clb.withFlags(ACC_FINAL + ACC_SUPER);
+                        clb.withInterfaceSymbols(List.of(CD_SYSTEM_MODULES));
+                        clb.withVersion(52, 0);
 
-            // generate hasSplitPackages
-            genHasSplitPackages(cw);
+                        // generate <init>
+                        genConstructor(clb);
 
-            // generate hasIncubatorModules
-            genIncubatorModules(cw);
+                        // generate hasSplitPackages
+                        genHasSplitPackages(clb);
 
-            // generate moduleDescriptors
-            genModuleDescriptorsMethod(cw);
+                        // generate hasIncubatorModules
+                        genIncubatorModules(clb);
 
-            // generate moduleTargets
-            genModuleTargetsMethod(cw);
+                        // generate moduleDescriptors
+                        genModuleDescriptorsMethod(clb);
 
-            // generate moduleHashes
-            genModuleHashesMethod(cw);
+                        // generate moduleTargets
+                        genModuleTargetsMethod(clb);
 
-            // generate moduleResolutions
-            genModuleResolutionsMethod(cw);
+                        // generate moduleHashes
+                        genModuleHashesMethod(clb);
 
-            // generate moduleReads
-            genModuleReads(cw, cf);
+                        // generate moduleResolutions
+                        genModuleResolutionsMethod(clb);
 
-            return cw;
+                        // generate moduleReads
+                        genModuleReads(clb, cf);
+                    });
         }
 
         /**
          * Generate bytecode for no-arg constructor
          */
-        private void genConstructor(ClassWriter cw) {
-            MethodVisitor mv = cw.visitMethod(ACC_PUBLIC, "<init>", "()V", null, null);
-            mv.visitVarInsn(ALOAD, 0);
-            mv.visitMethodInsn(INVOKESPECIAL,
-                               "java/lang/Object",
-                               "<init>",
-                               "()V",
-                               false);
-            mv.visitInsn(RETURN);
-            mv.visitMaxs(0, 0);
-            mv.visitEnd();
+        private void genConstructor(ClassBuilder clb) {
+            clb.withMethod("<init>", MethodTypeDesc.of(CD_void),
+                    ACC_PUBLIC, mb -> mb.withFlags(ACC_PUBLIC).withCode( cob -> {
+                        cob.loadInstruction(TypeKind.ReferenceType, 0);
+                        cob.invokeInstruction(Opcode.INVOKESPECIAL,
+                                              CD_Object,
+                                              "<init>",
+                                               MethodTypeDesc.of(CD_void),
+                                               false);
```

Comment:
nit: extra space 

sorry this might be from the patch I sent you.

Reply:
fixed, thanks.